### PR TITLE
Support NumPy 1.13 again

### LINKED ISSUE #342: numpy<1.13 do not have numpy.AxisError
When using numpy 1.12.1:
```python
>>> import cupy
>>> cupy.cumprod
<function cumprod at 0x7f9a460b4c80>
>>> cupy.cumprod(cupy.ndarray(()), axis=-10000)                                                                                                                                                             
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/path/to/cupy/math/sumprod.py", line 206, in cumprod
    raise numpy.AxisError('axis(={}) out of bounds'.format(axis))
AttributeError: module 'numpy' has no attribute 'AxisError'
```

### PR DESCRIPTION
**Pull request by okuta**
This PR includes #281, #336 and #344,
This PR fixes #342.

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -36,6 +36,17 @@ cdef inline _should_use_rop(x, y):
     return xp < yp and not isinstance(y, ndarray)
 
 
+class IndexOrValueError(IndexError, ValueError):
```
Comment (unnonouno) [inline cupy/core/core.pyx]: This error is not used with numpy 1.13, right?

Comment (okuta) [inline cupy/core/core.pyx]: Yes, If NumPy has `AxisError`, we use it.

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -36,6 +36,17 @@ cdef inline _should_use_rop(x, y):
...
+
+    def __init__(self, *args, **kwargs):
+        super(IndexOrValueError, self).__init__(*args, **kwargs)
+
+try:
+    AxisError = numpy.AxisError
+except AttributeError:
+    AxisError = IndexOrValueError
```
Comment (gwtnb) [inline cupy/core/core.pyx]: `cupy.core.core.AxisError` should not be used directly, so an underscore should be added.

**Comment by gwtnb** [comment]
LGTM except the above comment.

**Comment by okuta** [comment]
I fixed.

File: cupy/core/core.pyx (hunk 1 above)
Comment (unnonouno) [inline cupy/core/core.pyx]: If so, do we need to move it in `except` clause?

Comment (okuta) [inline cupy/core/core.pyx]: OK.

**Comment by gwtnb** [comment]
We can port this PR to v1 because this PR does not break exiting code. Users can still catch with `IndexError` or `ValueError`. [This behavior](https://github.com/cupy/cupy/pull/281/files#diff-d014fd086aae3978ecb209dbe84a0c23L43) is a bug.

**Comment by okuta** [comment]
> This behavior is a bug.
@gwtnb You mean this is bug in NumPy 1.13.0?

**Comment by gwtnb** [comment]
No, I mean it is a bug of `numpy<1.13` (or, at least an inconvenient behavior).

**Comment by okuta** [comment]
I understand.

File: cupy/core/__init__.py [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 
 # import class and function
+from cupy.core.core import _AxisError  # NOQA
```
Comment (gwtnb) [inline cupy/core/__init__.py]: Do we need this line?

Comment (okuta) [inline cupy/core/__init__.py]: OK, I delete this.

**Comment by gwtnb** [comment]
LGTM

**Pull request merged**